### DISCUSSION TITLE
Multiple custom cloud particles with Vue 2.X

### DISCUSSION THREAD

**Discussion by jasontheiler** [matteobruni/tsparticles#1260] [Q&A] [answered]
Hi,

I have problems understanding the documentation for this library. I'd like to create multiple custom cloud particles (for randomization), but don't know how to achieve that. In [this CodePen](https://codepen.io/matteobruni/pen/vYOMmGr) you use the `addShape` function to add a new custom shape, but I can't find detailed documentation for it and can't figure out how to do it with custom SVG paths. I don't even know how to get access to the `tsParticles` instance, that the Vue 2.X component uses under the hood.

I'd really appreciate any sort of help!

**Comment by matteobruni** [comment, marked as answer]
Hi @jasontheiler,

With the current stable release is not possible to customize the tsParticles instance like adding custom shapes, presets or other.
I noticed that a bit late and it will come out with the `2.0.0` version currently in alpha.

If you want to customize it I recommend to bypass the component using directly the core library. You can copy the component code ([here](https://github.com/matteobruni/tsparticles/blob/master/components/vue/src/Particles/Particles.vue)) for now customizing it

If you need more help creating a shape feel free to ask me any question. You can checkout some of the releases samples [here](https://github.com/matteobruni/tsparticles/tree/master/shapes)

**Reply by jasontheiler** [reply]
Hi @matteobruni,

thanks for the rapid reply!

Implementing the core library by itself worked so far. But the way that custom shapes are created seems kinda complicated to me. So I first tried using `type: "image"` with SVG files. That worked, sort of. But I couldn't override their `fill` color with `replaceColor: true`. They still had the original `fill` of the SVG. I'd really like to be able to change the colors via the options, so I don't have to load separate SVG files for a dark mode.

This is how the relevant bits of my options look like:
```js
import { tsParticles } from "tsparticles";

import cloud from "~/assets/graphics/cloud.svg";

...

particles: {
  color: {
    value: "#f00", // Should (in theory) color the SVGs red.
  },

  shape: {
    type: "image",
    options: {
      image: [
        {
          src: cloud,
          replaceColor: true,
        },
      ],
    },
  },
},

...

```

**Reply by matteobruni** [reply]
Glad to have helped @jasontheiler.

Can you share the SVG file? So I can test it on my own too. Maybe the SVG needs to be simplified.

Anyway, I have a question for you now. Custom shapes are shape types that doesn’t exist in the core library, `”image”` is one of the existing shapes, if you are gonna use only that you can use the Vue component.

**Reply by jasontheiler** [reply]
> Can you share the SVG file? So I can test it on my own too. Maybe the SVG needs to be simplified.

Sure! This is the SVG I currently try to use:
```html
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 272">
  <path
    d="M86,272C38.535,272 0,233.465 0,186C0,138.535 38.535,100 86,100C87.604,100 89.198,100.044 90.788,100.132C106.545,42.448 159.355,0 222,0C272.403,0 316.439,27.478 339.934,68.265C350.563,62.668 362.665,59.5 375.5,59.5C417.721,59.5 452,93.779 452,136C452,141.493 451.42,146.852 450.316,152.024C450.876,152.008 451.437,152 452,152C485.115,152 512,178.885 512,212C512,245.115 485.115,272 452,272L86,272Z"
    fill="#fff"
  />
</svg>

```

> Anyway, I have a question for you now. Custom shapes are shape types that doesn’t exist in the core library, `"image"` is one of the existing shapes, if you are gonna use only that you can use the Vue component.

I know that I wouldn't have to use the tsParticles core library by itself, if I don't add custom shapes. But since I want to use Nuxt's [color-mode](https://color-mode.nuxtjs.org/) module to switch between light/dark mode, I need to re-instanciate tsParticles every time someone switches between light/dark mode. Just using the tsParticles core library makes it easier to do that.

If you are curious, my current `Clouds.vue` component looks like this:
```vue
<template>
  <div :id="id" />
</template>

<script>
import { tsParticles } from "tsparticles";

import cloud from "~/assets/graphics/cloud.svg";

export default {
  props: {
    id: {
      type: String,
      required: true,
    },
  },

  data() {
    return {
      particlesContainer: undefined,
    };
  },

  watch: {
    "$colorMode.value"() {
      this.destroyParticles();
      this.loadParticles();
    },
  },

  mounted() {
    this.loadParticles();
  },

  beforeDestroy() {
    this.destroyParticles();
  },

  methods: {
    async loadParticles() {
      this.particlesContainer = await tsParticles.load(this.id, {
        detectRetina: true,
        particles: {
          move: {
            enable: true,
            direction: "bottom",
            straight: true,
            speed: 4,
          },

          color: {
            value: this.$colorMode.value === "dark" ? "#000" : "#fff", // Doesn't actually affect the particles when using `type: "image"`, but works with the default particle types.
          },

          shape: {
            type: "image",
            options: {
              image: [
                {
                  src: cloud,
                  replaceColor: true,
                },
              ],
            },
          },
        },
      });
    },

    destroyParticles() {
      this.particlesContainer?.destroy();
    },
  },
};
</script>

```

**Reply by jasontheiler** [reply]
Oh well, I just found the problem with the SVG files @matteobruni.

I have totally overlooked the fact that Nuxt automatically imports SVG files that are smaller than 1kB as a base64 data URL. tsParticles probably just can't handle the format and thus doesn't override its fills.

Since I also use the [Nuxt Optimized Images](https://marquez.co/docs/nuxt-optimized-images) module, I was able to fix it by importing the SVG with the query parameter `?url`. Now it works.

Example:
```js
// CommonJS
const cloud = require("~/assets/graphics/cloud.svg?url");

// ES6
import cloud from "~/assets/graphics/cloud.svg?url";
```

I'm still kinda interested in learning how to build custom shapes, though. It would reduce the amount of HTTP requests, if the particle shape wouldn't be an image.

Is there maybe a way to draw the custom shape with the path that is already defined in the SVG?

Anyway, thanks a ton!

**Reply by matteobruni** [reply]
> I'm still kinda interested in learning how to build custom shapes, though. It would reduce the amount of HTTP requests, if the particle shape wouldn't be an image.
> 
> Is there maybe a way to draw the custom shape with the path that is already defined in the SVG?

Actually not, custom shapes are just a function that takes the 2D context of the canvas, centered at the center of the particle and rotated correctly, to use any canvas drawing function.

You can see the system shapes here: https://github.com/matteobruni/tsparticles/tree/master/core/main/src/ShapeDrawers or the shapes I linked before.

The CodePen sample you linked is good too but it's very basic, at least it has some custom parameters.

[This](https://developer.mozilla.org/en-US/docs/Web/API/Canvas_API/Tutorial/Drawing_shapes) is a tutorial for drawing using the Canvas, you already have a `context` parameter ready to be used.

Another user requested a rounded rectangle shape but I'm focused on `2.0.0` version that will be a big release with great changes to the library. If you want to create a simple custom shape with a custom parameter (the border-radius) this could be a good starting point (Issue #1220).

If you start coding don't hesitate to ask for more help with that, I'm interested in new shapes, some interesting shapes could become part of the project that will be updated in the future releases

**Reply by jasontheiler** [reply]
I think I get the gist of it now, thanks!

I definitely wanna try creating a cloud shape (and maybe also a rounded square). But since I don't have a lot time right now, I'll probably stick with `type: "image"` first and revisit this topic in a few weeks, when things have settled down. If I have any questions or if I have some promising results, I'll definitely let you know!

Again, thanks a bunch!
